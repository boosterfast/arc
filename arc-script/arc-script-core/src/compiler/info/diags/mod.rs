pub mod repr;
pub mod to_codespan;
pub mod sink;
pub(crate) mod utils;

pub use repr::*;
pub use sink::*;
