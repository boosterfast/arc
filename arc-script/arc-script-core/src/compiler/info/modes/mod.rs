pub mod repr;
/// Module for representing sink and source information.
pub mod connector;

pub use repr::*;
