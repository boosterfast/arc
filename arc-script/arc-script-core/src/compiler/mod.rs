/// Module for representing Arc Queries.
pub(crate) mod query;
/// Module for representing the Abstract Syntax Tree.
///
/// NB: This module is public so that external libraries can construct ASTs
/// without having to generate source code.
pub mod ast;
/// Module for representing Dataflow graphs. Exports data types for staging
/// `HIR` functions.
pub mod dfg;
/// Module for representing Higher Order Intermediate Representation code.
pub(crate) mod hir;
/// Module for representing side information.
pub mod info;
/// Module for representing Multi-Level Intermediate Representation code.
pub(crate) mod mlir;
/// Module for representing Arcon code.
pub(crate) mod arcon;

// Module for incremental compilation.
// pub(crate) mod database;
/// Module which assembles compilation-pipeline.
pub mod pipeline;

pub(crate) mod pretty;

pub use pipeline::compile;
