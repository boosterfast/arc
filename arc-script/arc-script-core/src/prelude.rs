//! Exports for users who use the compiler as a library.

pub use crate::compiler;
pub use crate::compiler::ast;
pub use crate::compiler::info::diags;
pub use crate::compiler::info::logger;
pub use crate::compiler::info::modes;

pub use anyhow::Result;
pub use half::bf16;
pub use half::f16;
