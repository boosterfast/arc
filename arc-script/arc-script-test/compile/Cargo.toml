[package]
name        = "arc-script-test-compile"
version     = "0.0.0"
authors     = ["Klas Segeljakt <klasseg@kth.se>"]
edition     = "2018"
license     = "MIT"
description = "Tests to see if arc-scripts compile"

[dependencies]
trybuild        = { version = "1.0.39" }
arc-script-core = { version = "=0.0.0", path = "../../arc-script-core" }
insta           = { version = "1.6.0", features = ["glob"] }

[dev-dependencies]
arcon           = { git = "https://github.com/cda-group/arcon" }
arc-script      = { version = "=0.0.0", path = "../../" }
prost           = { version = "0.7.0" }
