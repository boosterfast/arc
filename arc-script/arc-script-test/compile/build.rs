fn main() {
    arc_script_build::Builder::default()
        .no_exclude(true)
        .source_dirs(["src/tests/expect_pass", "src/tests/expect_mlir_fail_todo"])
        .build();
}
