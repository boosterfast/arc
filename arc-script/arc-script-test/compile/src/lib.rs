#![allow(unused)]

#[cfg(test)]
mod insta;

// #[cfg(test)]
// mod trybuild;

// #[cfg(test)]
// mod compiletest;

mod tests {
    mod expect_pass {
        mod basic_pipe;
        mod binops;
        mod enum_pattern;
        mod enum_pattern_nested;
        mod enums;
        mod fib;
        mod fun;
        mod if_let;
        mod ifs;
        mod lambda;
        mod literals;
        mod nested_if;
        mod option;
        mod path;
        mod pattern;
        mod r#if;
        mod structs;
    }
    mod expect_mlir_fail_todo {
        mod extern_fun;
        mod pipe;
        mod stateful;
        mod task_filter;
        mod task_identity_untagged;
        mod task_map;
        mod task_with_funs;
    }
}
