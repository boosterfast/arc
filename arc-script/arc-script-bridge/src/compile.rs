use arc_script_core::compiler::compile;
use arc_script_core::compiler::info::diags::sink::Buffer;
use arc_script_core::prelude::modes::{Input, Mode, Output};

use proc_macro as pm;
use quote::quote;

use std::fs;
use std::path::PathBuf;

/// arc_script::compile!("script.arc")
///    .stage_fun1(1, 2, 3)
///    .stage_fun2("foo", "bar")
///    .finalize()
pub(crate) fn expand(input: pm::TokenStream) -> pm::TokenStream {
    todo!()
}
