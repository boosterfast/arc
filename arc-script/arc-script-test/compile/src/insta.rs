#![cfg_attr(not(test), allow(dead_code, unused_imports))]

use arc_script_core::prelude::compiler::compile;
use arc_script_core::prelude::diags::Buffer;
use arc_script_core::prelude::modes::{Input, Mode, Output};

use std::env;

fn snapshot(output: Output, paths: &str) {
    env::set_var("INSTA_OUTPUT", "summary");
    env::set_var("INSTA_FORCE_PASS", "1");
    let mut settings = insta::Settings::clone_current();
    settings.set_prepend_module_to_snapshot(false);
    settings.remove_snapshot_suffix();
    let mut sink = Buffer::no_color();
    settings.bind(|| {
        insta::glob!(paths, |path| {
            println!("Testing {}", path.display());
            let mode = Mode {
                input: Input::File(Some(path.into())),
                output,
                ..Default::default()
            };
            compile(mode, &mut sink).unwrap();
            let s = std::str::from_utf8(sink.as_slice()).unwrap();
            insta::assert_snapshot!(s);
            sink.clear();
        })
    });
}

#[test]
fn test_ast() {
    snapshot(Output::AST, "tests/expect_mlir_fail_todo/*.arc");
    snapshot(Output::AST, "tests/expect_pass/*.arc");
    snapshot(Output::AST, "tests/expect_fail/*.arc");
    snapshot(Output::AST, "tests/expect_fail_todo/*.arc");
}

#[test]
fn test_hir() {
    snapshot(Output::HIR, "tests/expect_mlir_fail_todo/*.arc");
    snapshot(Output::HIR, "tests/expect_pass/*.arc");
    snapshot(Output::HIR, "tests/expect_fail/*.arc");
    snapshot(Output::HIR, "tests/expect_fail_todo/*.arc");
}

#[test]
fn test_dfg() {
    snapshot(Output::DFG, "tests/expect_mlir_fail_todo/*.arc");
    snapshot(Output::DFG, "tests/expect_pass/*.arc");
    snapshot(Output::DFG, "tests/expect_fail/*.arc");
    snapshot(Output::DFG, "tests/expect_fail_todo/*.arc");
}

#[test]
fn test_rust() {
    snapshot(Output::Rust, "tests/expect_mlir_fail_todo/*.arc");
    snapshot(Output::Rust, "tests/expect_pass/*.arc");
    snapshot(Output::Rust, "tests/expect_fail/*.arc");
    snapshot(Output::Rust, "tests/expect_fail_todo/*.arc");
}

#[test]
fn test_mlir() {
    snapshot(Output::MLIR, "tests/expect_pass/*.arc");
    snapshot(Output::MLIR, "tests/expect_fail/*.arc");
    snapshot(Output::MLIR, "tests/expect_fail_todo/*.arc");
}
