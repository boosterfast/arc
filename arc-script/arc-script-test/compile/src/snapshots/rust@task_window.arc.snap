---
source: arc-script-test/compile/src/insta.rs
expression: s
input_file: arc-script-test/compile/src/tests/expect_mlir_fail_todo/task_window.arc

---
#[allow(non_snake_case)]
#[allow(unused_must_use)]
#[allow(dead_code)]
#[allow(unused_variables)]
#[allow(unused_imports)]
#[allow(unused_braces)]
pub mod defs {
    use super::*;
    pub use arc_script::arcorn;
    pub use arc_script::arcorn::state::{ArcMapOps, ArcRefOps, ArcSetOps, ArcVecOps};
    pub use arcon::prelude::*;
    pub struct TumblingWindowSum<'i, 'source, 'timer, 'channel, B: Backend, C: ComponentDefinition> {
        pub data: &'i mut TumblingWindowSumData,
        pub ctx: &'i mut OperatorContext<'source, 'timer, 'channel, TumblingWindowSumData, B, C>,
        pub timestamp: Option<u64>,
        pub key: u64,
    }
    pub struct TumblingWindowSumData {
        pub state: TumblingWindowSumState,
    }
    #[derive(ArconState)]
    pub struct TumblingWindowSumState {
        pub agg_0: arc_script::arcorn::state::ArcRef<i32, Sled>,
    }
    impl StateConstructor for TumblingWindowSumState {
        type BackendType = Sled;
        fn new(backend: Arc<Self::BackendType>) -> Self {
            Self {
                agg_0: <arc_script::arcorn::state::ArcRef<i32, Sled>>::arc_ref_default(
                    "agg_0",
                    backend.clone(),
                )
                .unwrap(),
            }
        }
    }
    impl TumblingWindowSumData {
        fn new() -> OperatorBuilder<TumblingWindowSumData> {
            OperatorBuilder {
                constructor: Arc::new(move |backend| TumblingWindowSumData {
                    state: TumblingWindowSumState {
                        agg_0: <arc_script::arcorn::state::ArcRef<i32, Sled>>::arc_ref_new(
                            "agg_0",
                            backend.clone(),
                            {
                                let y_1_0 = 0i32;
                                y_1_0
                            },
                        )
                        .unwrap(),
                    },
                }),
                conf: Default::default(),
            }
        }
    }
    impl Operator for TumblingWindowSumData {
        type IN = Struct3keyi323vali32End;
        type OUT = Struct3keyi323vali32End;
        type TimerState = Struct3keyi323valUnitEnd;
        type OperatorState = TumblingWindowSumState;
        fn handle_element(
            &mut self,
            element: ArconElement<Self::IN>,
            ref mut ctx: OperatorContext<Self, impl Backend, impl ComponentDefinition>,
        ) -> OperatorResult<()> {
            let ArconElement { timestamp, data } = element;
            let mut task = TumblingWindowSum {
                data: self,
                ctx,
                timestamp,
                key: data.get_key(),
            };
            task.handle_element(data);
            Ok(())
        }
        fn state(&mut self) -> &mut Self::OperatorState {
            &mut self.state
        }
        fn handle_timeout(
            &mut self,
            timeout: Self::TimerState,
            ref mut ctx: OperatorContext<Self, impl Backend, impl ComponentDefinition>,
        ) -> OperatorResult<()> {
            let mut task = TumblingWindowSum {
                data: self,
                key: timeout.get_key(),
                ctx,
                timestamp: None,
            };
            task.handle_timeout(timeout);
            Ok(())
        }
        fn persist(&mut self) -> OperatorResult<()> {
            self.state.persist();
            Ok(())
        }
    }
    impl<'i, 'source, 'timer, 'channel, B: Backend, C: ComponentDefinition>
        TumblingWindowSum<'i, 'source, 'timer, 'channel, B, C>
    {
        fn handle_element(&mut self, x_0: Struct3keyi323vali32End) -> OperatorResult<()> {
            {
                let y_1_0 = x_0.val;
                let y_1_1 = x_0.key;
                let y_1_2 = self.data.state.agg_0.arc_ref_read().unwrap();
                let y_1_3 = y_1_2 + y_1_0;
                self.data.state.agg_0.arc_ref_write(y_1_3).unwrap()
            };
            Ok(())
        }
        fn emit(&mut self, data: Struct3keyi323vali32End) {
            let element = ArconElement {
                data,
                timestamp: self.timestamp,
            };
            self.ctx.output(element);
        }
        fn handle_timeout(&mut self, x_1: Struct3keyi323valUnitEnd) {
            {
                let y_1_0 = x_1.val;
                let y_1_1 = ();
                let y_1_2 = y_1_0 == y_1_1;
                let y_1_3 = if y_1_2 {
                    let y_2_0 = x_1.key;
                    let y_2_1 = self.data.state.agg_0.arc_ref_read().unwrap();
                    let y_2_2 = Struct3keyi323vali32End {
                        val: y_2_1,
                        key: y_2_0,
                    };
                    let y_2_3 = 0i32;
                    let y_2_4 = ();
                    let y_2_5 = Struct3keyi323valUnitEnd {
                        val: y_2_4,
                        key: y_2_0,
                    };
                    let y_2_6 = 60u64;
                    let y_2_7 = Struct3durduration3valStruct3keyi323valUnitEndEnd {
                        val: y_2_5,
                        dur: y_2_6,
                    };
                    let y_2_8 = self.data.state.agg_0.arc_ref_write(y_2_3).unwrap();
                    self.trigger(y_2_7.val, y_2_7.dur);
                    let y_2_9 = self.emit(y_2_2);
                    y_2_8;
                    y_2_9
                } else {
                    let y_2_0 = todo!();
                    y_2_0
                };
                y_1_3
            }
        }
        fn trigger(&mut self, data: Struct3keyi323valUnitEnd, after: u64) {
            let time = self.ctx.current_time().unwrap() + after;
            self.ctx.schedule_at(self.key, time, data);
        }
    }
    #[arcorn::rewrite]
    #[derive(Copy)]
    pub struct Struct3keyi323valUnitEnd {
        key: i32,
        val: (),
    }
    #[arcorn::rewrite]
    #[derive(Copy)]
    pub struct Struct3durduration3valStruct3keyi323valUnitEndEnd {
        dur: u64,
        val: Struct3keyi323valUnitEnd,
    }
    #[arcorn::rewrite]
    #[derive(Copy)]
    pub struct Struct3keyi323vali32End {
        key: i32,
        val: i32,
    }
}
pub use defs::*;


